### PR TITLE
[DEV-4798] template repositories

### PR DESCRIPTION
**Pull request by gaetano-guerriero**

### CODE REVIEW

File: openapi.yaml [hunk 1]
```diff
@@ -3504,13 +3504,25 @@ components:
...
       type: object
+    GoalRepositories:
+# {% if False %}
+      allOf:
+# {% else %}{{ "\n" }}      anyOf:{% endif %}
+      - $ref: '#/components/schemas/RepositorySet'
+      - items:
+          type: string
+        nullable: true
+        type: array
```
Comment (vmarkovtsev) [inline openapi.yaml:3515]: No ambiguous types, please. Always require a list, no scalar values.

File: openapi.yaml [hunk 2]
```diff
@@ -3524,10 +3536,13 @@ components:
           - $ref: '#/components/schemas/ReleaseMetricID'
         name:
           $ref: '#/components/schemas/GoalTemplateName'
+        repositories:
+          $ref: '#/components/schemas/GoalRepositories'
       required:
       - id
       - name
       - metric
+      - repositories
```
Comment (vmarkovtsev) [inline openapi.yaml:3545]: It shouldn't be required.

File: openapi.yaml [hunk 3]
```diff
@@ -3536,6 +3551,7 @@ components:
         account: 30
         metric: pr-review-time
         name: My New Goal Template
+        repositories: null
```
Comment (vmarkovtsev) [inline openapi.yaml:3554]: Hence, we can remove this.

File: openapi.yaml [hunk 4]
```diff
@@ -3549,20 +3565,26 @@ components:
           - $ref: '#/components/schemas/ReleaseMetricID'
         name:
           $ref: '#/components/schemas/GoalTemplateName'
+        repositories:
+          $ref: '#/components/schemas/GoalRepositories'
       required:
       - account
       - metric
       - name
+      - repositories
```
Comment (vmarkovtsev) [inline openapi.yaml:3574]: Not required here, too.

File: openapi.yaml [hunk 5]
```diff
@@ -3576,11 +3598,15 @@ components:
...
         name: New template name
+        repositories: null
       properties:
         name:
           $ref: '#/components/schemas/GoalTemplateName'
+        repositories:
+          $ref: '#/components/schemas/GoalRepositories'
       required:
       - name
+      - repositories
```
Comment (vmarkovtsev) [inline openapi.yaml:3609]: Not required.

File: openapi.yaml [hunk 6]
```diff
@@ -3576,11 +3598,15 @@ components:
       description: Goal template creation request.
       example:
         name: New template name
+        repositories: null
```
Comment (vmarkovtsev) [inline openapi.yaml:3601]: Removed.

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: openapi.yaml (hunk 2 above)
Comment (gaetano-guerriero) [reply openapi.yaml:3545]: You mean that the API is not required to include `repositories` in the response ?

File: openapi.yaml (hunk 5 above)
Comment (gaetano-guerriero) [reply openapi.yaml:3609]: What is the semantic of not including `repositories` in the `update_goal_template` operation ?
The API should set the field to `null` ? Or the API should implement a partial update logic and not change the field ?
Do you have other examples of partial updates in the API ?

File: openapi.yaml (hunk 1 above)
Comment (gaetano-guerriero) [reply openapi.yaml:3515]: Sorry I don't understand, valid values are
- `null`
- `[]`
- `["github.com/a"]

File: openapi.yaml (hunk 5 above)
Comment (gaetano-guerriero) [reply openapi.yaml:3609]: I guess on creation it can be optional and be interpreted as `null` instead

File: openapi.yaml (hunk 1 above)
Comment (vmarkovtsev) [reply openapi.yaml]: No `null` should be allowed. Either a list or not sent at all. This is consistent with how the rest of the models already work.

File: openapi.yaml (hunk 2 above)
Comment (vmarkovtsev) [reply openapi.yaml:3541]: Yep

File: openapi.yaml (hunk 1 above)
Comment (gaetano-guerriero) [reply openapi.yaml]: So `[]` means "the goals refers to every repositories" ?

File: openapi.yaml (hunk 5 above)
Comment (vmarkovtsev) [reply openapi.yaml:3603]: The request doesn't contain the fields that shouldn't be updated. This is how it's working for other endpoints and models. Also, we have a PUT that overwrites the whole goal, so we don't support partial updates. We don't support partial updates anywhere in the current endpoints (well, except maybe `PUT /account/{id}/features`, but it's arguable), and it's a good thing.

File: openapi.yaml (hunk 2 above)
Comment (gaetano-guerriero) [reply openapi.yaml]: fixed

**Review by vmarkovtsev — APPROVED** [review]

**Pull request merged**